### PR TITLE
Add AES128-SHA256 and ECDHE-RSA-AES128-SHA256 ciphers

### PR DESCRIPTION
**Pull request by dkostic**
### Issues:
CryptoAlg-1065

### Description of changes: 
The change adds support for two ciphers:
- `AES128-SHA256` (`TLS_RSA_WITH_AES_128_CBC_SHA256`)
- `ECDHE-RSA-AES128-SHA256` (`TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256`)

### Testing:
`./crypto/crypto_test, ./ssl/ssl_test, ./ssl/test/runner`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/openssl/aead.h [hunk 1]
```diff
@@ -402,6 +402,12 @@ OPENSSL_EXPORT const EVP_AEAD *EVP_aead_aes_128_cbc_sha1_tls_implicit_iv(void);
 OPENSSL_EXPORT const EVP_AEAD *EVP_aead_aes_256_cbc_sha1_tls(void);
 OPENSSL_EXPORT const EVP_AEAD *EVP_aead_aes_256_cbc_sha1_tls_implicit_iv(void);
 
+OPENSSL_EXPORT const EVP_AEAD *EVP_aead_aes_128_cbc_sha256_tls(void);
+OPENSSL_EXPORT const EVP_AEAD *EVP_aead_aes_128_cbc_sha256_tls_implicit_iv(void);
+
+OPENSSL_EXPORT const EVP_AEAD *EVP_aead_aes_256_cbc_sha256_tls(void);
+OPENSSL_EXPORT const EVP_AEAD *EVP_aead_aes_256_cbc_sha256_tls_implicit_iv(void);
+
```
Comment (bryce-shang) [inline include/openssl/aead.h:410]: Need AEAD cipher tests?

Example: https://github.com/awslabs/aws-lc/blob/725625435158150ef21e0a4dab6fa3aca1ef2d2c/crypto/cipher_extra/aead_test.cc#L128-L135

How to generate the test
https://github.com/awslabs/aws-lc/pull/442

File: crypto/cipher_extra/tls_cbc.c [hunk 2]
```diff
@@ -336,3 +333,189 @@ int EVP_tls_cbc_digest_record(const EVP_MD *md, uint8_t *md_out,
...
+  size_t total_bits = ctx->Nl + (len << 3);
+  uint8_t length_bytes[4];
+  length_bytes[0] = (uint8_t)(total_bits >> 24);
+  length_bytes[1] = (uint8_t)(total_bits >> 16);
+  length_bytes[2] = (uint8_t)(total_bits >> 8);
+  length_bytes[3] = (uint8_t)total_bits;
+
+  // We now construct and process each expected block in constant-time.
+  uint8_t block[SHA256_CBLOCK] = {0};
+  uint32_t result[8] = {0};
```
Comment (torben-hansen) [inline crypto/cipher_extra/tls_cbc.c:373]: magic number - also pointing at myself :)
It's the size of the state counted in 32-bit numbers.

File: crypto/cipher_extra/tls_cbc.c [hunk 3]
```diff
@@ -336,3 +333,189 @@ int EVP_tls_cbc_digest_record(const EVP_MD *md, uint8_t *md_out,
...
+
+    // Fill in the length if this is the last block.
+    crypto_word_t is_last_block = constant_time_eq_w(i, last_block);
+    for (size_t j = 0; j < 4; j++) {
+      block[SHA256_CBLOCK - 4 + j] |= is_last_block & length_bytes[j];
+    }
+
+    // Process the block and save the hash state if it is the final value.
+    SHA256_Transform(ctx, block);
+    for (size_t j = 0; j < 8; j++) {
```
Comment (torben-hansen) [inline crypto/cipher_extra/tls_cbc.c:418]: Similar magic numbers here and below.

File: crypto/cipher_extra/tls_cbc.c [hunk 4]
```diff
@@ -336,3 +333,189 @@ int EVP_tls_cbc_digest_record(const EVP_MD *md, uint8_t *md_out,
...
+  // Write the output.
+  for (size_t i = 0; i < 8; i++) {
+    CRYPTO_store_u32_be(out + 4 * i, result[i]);
+  }
+  return 1;
+}
+
+static int EVP_tls_cbc_digest_record_sha256(
+              const EVP_MD *md, uint8_t *md_out,
+              size_t *md_out_size, const uint8_t header[13],
```
Comment (torben-hansen) [inline crypto/cipher_extra/tls_cbc.c:432]: magic number.

File: crypto/cipher_extra/tls_cbc.c [hunk 5]
```diff
@@ -336,3 +333,189 @@ int EVP_tls_cbc_digest_record(const EVP_MD *md, uint8_t *md_out,
   *md_out_size = SHA_DIGEST_LENGTH;
   return 1;
 }
+
+int EVP_sha256_final_with_secret_suffix(SHA256_CTX *ctx,
```
Comment (torben-hansen) [inline crypto/cipher_extra/tls_cbc.c:337]: Should probably had aligned with the name for `EVP_tls_cbc_digest_record_sha256()`.

File: ssl/ssl_test.cc [hunk 6]
```diff
@@ -176,11 +176,13 @@ static const CipherTest kCipherTests[] = {
         "ECDHE-ECDSA-CHACHA20-POLY1305:"
         "ECDHE-RSA-CHACHA20-POLY1305:"
         "ECDHE-ECDSA-AES128-GCM-SHA256:"
+        "ECDHE-RSA-AES128-SHA256:"
```
Comment (torben-hansen) [inline ssl/ssl_test.cc:179]: Can probably come up with a million different tests here; are we happy with the current ones+the ones added in this PR?

File: crypto/cipher_extra/cipher_test.cc [hunk 7]
```diff
@@ -525,6 +525,53 @@ TEST(CipherTest, SHA1WithSecretSuffix) {
   }
 }
 
+TEST(CipherTest, SHA256WithSecretSuffix) {
```
Comment (torben-hansen) [inline crypto/cipher_extra/cipher_test.cc:528]: Is this test actually functional (i.e. does it do any CT checking)?

**Review by torben-hansen — COMMENTED** [review]
lgtm. To finish should:
* Run some timing experiments
* As Bryce points out, get same functional test vectors as we have for sha1-variants.

File: ssl/ssl_test.cc (hunk 6 above)
Comment (dkostic) [reply ssl/ssl_test.cc:179]: I think we are fine, but I can try to come up with some meaningful new tests.

File: ssl/ssl_cipher.cc [hunk 8]
```diff
@@ -627,6 +651,26 @@ bool ssl_cipher_get_evp_aead(const EVP_AEAD **out_aead,
...
+      } else {
+        *out_aead = EVP_aead_aes_128_cbc_sha256_tls();
+      }
+    } else if (cipher->algorithm_enc == SSL_AES256) {
+      if (version == TLS1_VERSION) {
+        *out_aead = EVP_aead_aes_256_cbc_sha256_tls_implicit_iv();
+        *out_fixed_iv_len = 16;
+      } else {
+        *out_aead = EVP_aead_aes_256_cbc_sha256_tls();
+      }
```
Comment (bryce-shang) [inline ssl/ssl_cipher.cc:668]: Why `AES_256_CBC_SHA256` is configured? `CryptoAlg-1065` only requests `AES_128_CBC_SHA256`.

File: ssl/ssl_test.cc [hunk 9]
```diff
@@ -1034,6 +1038,15 @@ TEST(SSLTest, CipherProperties) {
           NID_auth_rsa,
           NID_md5_sha1,
       },
+      {
```
Comment (bryce-shang) [inline ssl/ssl_test.cc:1041]: Upstream removed the two ciphers in [this review](https://boringssl-review.googlesource.com/c/boringssl/+/27944/). The removed diffs of `ssl_test.cc` include some tests we may be interested in.

File: crypto/cipher_extra/cipher_test.cc (hunk 7 above)
Comment (bryce-shang) [reply crypto/cipher_extra/cipher_test.cc:528]: `SHA256WithSecretSuffix` CT was checked in `CryptoAlg-1067`. Basically, some constant code is changed to non constant and the valgrind can detect the change. e.g.
```
[ RUN      ] CipherTest.SHA256WithSecretSuffix
==3891== Conditional jump or move depends on uninitialised value(s)
==3891==    at 0x566BA6: constant_time_is_zero_w (internal.h:388)
==3891==    by 0x566BD7: constant_time_eq_w (internal.h:405)
==3891==    by 0x566BFC: constant_time_eq_8 (internal.h:411)
```

File: ssl/ssl_cipher.cc [hunk 10]
```diff
@@ -627,6 +651,26 @@ bool ssl_cipher_get_evp_aead(const EVP_AEAD **out_aead,
...
     *out_mac_secret_len = SHA_DIGEST_LENGTH;
+  } else if (cipher->algorithm_mac == SSL_SHA256) {
+    if (cipher->algorithm_enc == SSL_AES128) {
+      if (version == TLS1_VERSION) {
+        *out_aead = EVP_aead_aes_128_cbc_sha256_tls_implicit_iv();
+        *out_fixed_iv_len = 16;
+      } else {
+        *out_aead = EVP_aead_aes_128_cbc_sha256_tls();
+      }
+    } else if (cipher->algorithm_enc == SSL_AES256) {
```
Comment (torben-hansen) [inline ssl/ssl_cipher.cc:662]: None of the cipher suites in the cipher suite list specify `CBC` + `aes256` + `sha256`. So, this looks like dead code.

File: crypto/cipher_extra/tls_cbc.c (hunk 4 above)
Comment (dkostic) [reply crypto/cipher_extra/tls_cbc.c:432]: done

File: ssl/ssl_test.cc (hunk 9 above)
Comment (dkostic) [reply ssl/ssl_test.cc:1041]: thanks Bryce, I added two rules based on the ones removed in the quoted upstream change.

File: crypto/cipher_extra/tls_cbc.c (hunk 5 above)
Comment (dkostic) [reply crypto/cipher_extra/tls_cbc.c]: done

**Review by bryce-shang — APPROVED** [review]

**Review by torben-hansen — APPROVED** [review]

**Review by torben-hansen — APPROVED** [review]

**Review by andrewhop — APPROVED** [review]

**Pull request merged**